Minimal Bubbling Observers

NthTensor opened the pull request:
# Objective

Add basic bubbling to observers, modeled off `bevy_eventlistener`.

## Solution

- Introduce a new `Traversal` trait for components which point to other entities.
- Provide a default `TraverseNone: Traversal` component which cannot be constructed.
- Implement `Traversal` for `Parent`.
- The `Event` trait now has an associated `Traversal` which defaults to `TraverseNone`.
- Added a field `bubbling: &mut bool` to `Trigger` which can be used to instruct the runner to bubble the event to the entity specified by the event's traversal type.
- Added an associated constant `SHOULD_BUBBLE` to `Event` which configures the default bubbling state.
- Added logic to wire this all up correctly.

Introducing the new associated information directly on `Event` (instead of a new `BubblingEvent` trait) lets us dispatch both bubbling and non-bubbling events through the same api.

## Testing

I have added several unit tests to cover the common bugs I identified during development. Running the unit tests should be enough to validate correctness. The changes effect unsafe portions of the code, but should not change any of the safety assertions. 

## Changelog

Observers can now bubble up the entity hierarchy! To create a bubbling event, change your `Derive(Event)` to something like the following:

```rust
#[derive(Component)]
struct MyEvent;

impl Event for MyEvent {
    type Traverse = Parent; // This event will propagate up from child to parent.
    const AUTO_PROPAGATE: bool = true; // This event will propagate by default.
}
```

You can dispatch a bubbling event using the normal `world.trigger_targets(MyEvent, entity)`. 

Halting an event mid-bubble can be done using `trigger.propagate(false)`. Events with `AUTO_PROPAGATE = false` will not propagate by default, but you can enable it using `trigger.propagate(true)`.

If there are multiple observers attached to a target, they will all be triggered by bubbling. They all share a bubbling state, which can be accessed mutably using `trigger.propagation_mut()` (`trigger.propagate` is just sugar for this). 

You can choose to implement `Traversal` for your own types, if you want to bubble along a different structure than provided by `bevy_hierarchy`. Implementers must be careful never to produce loops, because this will cause bevy to hang. 

## Migration Guide
+ Manual implementations of `Event` should add associated type `Traverse = TraverseNone` and associated constant `AUTO_PROPAGATE = false`;
+ `Trigger::new` has new field `propagation: &mut Propagation` which provides the bubbling state.
+ `ObserverRunner` now takes the same `&mut Propagation` as a final parameter.

MrGVSV commented on crates/bevy_ecs/src/observer/mod.rs:
```diff
@@ -56,6 +59,21 @@ impl<'w, E, B: Bundle> Trigger<'w, E, B> {
         self.trigger.entity
     }
+
+    /// Enables event bubbling.
+    pub fn propagate(&mut self, should_bubble: bool) {
+        *self.bubble = should_bubble;
+    }
```
I could be wrong but I think "bubbling" is a subset of "propagation" (specifically one that goes _up_ a hierarchy). It might be better to keep these two terms separate from one another, and have the parameter be `should_propagate`.

Or we could also consider moving away from the web-based terminology altogether and call `bubbles`/`should_bubble` something like `traverses`/`should_traverse`?

NthTensor replied:
Good point. I'm good with whatever people like best. The terminology here is ported pretty directly from `bevy_eventlistener` which mixes the two.

alice-i-cecile replied:
I prefer propagation for the generic term, and bubbling for the "walk up the hierarchy" behavior specifically.

MrGVSV replied:
> I prefer propagation for the generic term, and bubbling for the "walk up the hierarchy" behavior specifically.

One potential issue is that we don't control which way we're walking. If a library implements `Event` where `Traverse` is `MainChild` or something, then we're technically walking down the tree.

This might get even more confusing if we introduce a second phase to the propagation (like a "capture" phase). It might happen that the way a particular event is set up that things are seemingly swapped to the consumer of that event.

Not sure how likely this all is but worth thinking about imo.

Maybe we remove the term "bubble" entirely and prefer "traverse" in its place? Or maybe we rename the assoc type to `Bubble` to indicate the intended direction (I guess that's assuming we ever plan on adding a second phase). Or maybe it's not that big a deal and we just leave it as is haha

NthTensor replied:
> In any case, my point is don't spend any extra effort enabling a use case which might never happen.

I agree with this. I'd be ok actually going the other way and making the `Traversal` trait more specific to upward propagation/bubbling. The relations working group have told me there's little chance this can be forward compatible with relations, so there's no point trying to broaden the scope too much.

MrGVSV replied:
> > In any case, my point is don't spend any extra effort enabling a use case which might never happen.
> 
> 
> 
> I agree with this. I'd be ok actually going the other way and making the `Traversal` trait more specific to upward propagation/bubbling.

I could be completely misinterpreting this but I feel like naming it something more specific to the directionality is more for the purpose of enabling future behavior in alternative directions. If we want to avoid the issue altogether then I feel like a general `traverse`/`propagate` term might be clearer in that case rather than doubling on the bubbling (sorry lol)

NthTensor replied:
I am fine with any of it but I really would like to get this cleaned up for approval as soon as possible. @alice-i-cecile how do we feel about keeping the trait as "traversal" and changing the observers/events part to always use "bubbling" over "propagation"?

alice-i-cecile replied:
Sounds fine 👍🏽

NthTensor replied:
Ended up going with Propagation everywhere, since `Bubbling` is more of a `bevy_hierarchy` concept.

MrGVSV commented on crates/bevy_ecs/src/observer/mod.rs:
```diff
@@ -662,4 +755,162 @@ mod tests {
     }
+
+    #[test]
+    fn observer_bubling() {
```
```suggestion
    fn observer_bubbling() {
```

MrGVSV commented on crates/bevy_ecs/src/observer/mod.rs:
```diff
@@ -662,4 +755,162 @@ mod tests {
+    }
+
+    #[test]
+    fn observer_bubling_redundant_dispatch() {
```
```suggestion
    fn observer_bubbling_redundant_dispatch() {
```

MrGVSV commented on crates/bevy_ecs/src/observer/mod.rs:
```diff
@@ -662,4 +755,162 @@ mod tests {
+    }
+
+    #[test]
+    fn observer_bubling_stop_propagation() {
```
```suggestion
    fn observer_bubbling_stop_propagation() {
```

MrGVSV commented on crates/bevy_ecs/src/observer/mod.rs:
```diff
@@ -662,4 +755,162 @@ mod tests {
+    }
+
+    #[test]
+    fn observer_bubling_join() {
```
```suggestion
    fn observer_bubbling_join() {
```

MrGVSV commented on crates/bevy_ecs/src/observer/mod.rs:
```diff
@@ -662,4 +755,162 @@ mod tests {
+    }
+
+    #[test]
+    fn observer_bubling_no_next() {
```
```suggestion
    fn observer_bubbling_no_next() {
```

MrGVSV commented on crates/bevy_ecs/src/observer/mod.rs:
```diff
@@ -662,4 +755,162 @@ mod tests {
+    }
+
+    #[test]
+    fn observer_bubling_parallel_propagation() {
```
```suggestion
    fn observer_bubbling_parallel_propagation() {
```

MrGVSV commented on crates/bevy_ecs/src/observer/mod.rs:
```diff
@@ -421,6 +496,24 @@ mod tests {
+    }
+
+    #[derive(Component)]
+    struct EventBubling;
```
Typo: should be `EventBubbling`

MrGVSV reviewed (commented):
Just a few typos on some internal test items (awesome tests btw!)

Jondolf reviewed (commented):
I'm not too familiar with the observer code (or ECS details in general), but I didn't spot any obvious issues based on a quick look. I mainly went through the docs and how this works at a high level.

The core implementation looks solid, and will definitely be highly useful for picking and UI, as well as many general patterns. I like that you can implement custom traversal: for example in `bevy_xpbd`, I have a `ColliderParent` component, and I could use it to propagate collision events from child colliders directly to rigid body entities.

I do agree with @MrGVSV that the current terminology is potentially misleading. Bubbling in my mind always goes *up* in the hierarchy, which is not necessarily the case for us if we allow custom traversal.

Overall I'd like to see some improvements to the documentation here: What is bubbling, and how do we define it? What could it be useful for? How do I stop propagation? What is `Event::Traverse` actually used for? How can I implement my own traversal? These are largely answered in the PR description, but I'd like it in the actual documentation.

NthTensor commented:
I would like advice on how best to document this, since it doesn't really expose anything publicly by default, except for perhaps the new `Traversal` trait which I am writing better docs for.

The docs entry-point for observers seems to be the `Observer` component. Does it make more sense to talk about this there, on `Event` or both?

alice-i-cecile commented:
My preference is to primarily cover this on `Traversal`, but link to that trait's docs with a one sentence explanation of how / why this might be used in both `Event` and `Observer`'s docs.

NthTensor commented on crates/bevy_ecs/src/traversal.rs:
```diff
@@ -0,0 +1,33 @@
+/// The implementor is responsible for ensuring that `Traversal::next` cannot produce infinite loops.
+pub trait Traversal: Component {
+    /// Returns the next entity to visit.
+    fn next(&self) -> Option<Entity>;
```
I experimented with several versions of this api and ultimately decided to implement the minimum necessary for the picking port.

Allowing multiple outputs, either with a `&mut Self` or a `&[Entity]` ended up significantly complicating observer dispatch because it requires a branching stack of dispatched events, where traversal could be halted at any point.

Which is to say, I'd really like to leave that sort of thing to follow-up.

re0312 commented on crates/bevy_ecs/src/observer/mod.rs:
```diff
@@ -197,30 +219,82 @@ impl Observers {
+                                break;
+                            }
+                        }
+                        if let Some(next) = world.get_mut::<C>(entity).and_then(|t| t.next()) {
```
```suggestion
                        if let Some(next) = world.get::<C>(entity).and_then(|t| t.next()) {
```

re0312 commented on crates/bevy_ecs/src/observer/mod.rs:
```diff
@@ -197,30 +219,82 @@ impl Observers {
+                        break;
+                    }
+                }
+                if let Some(next) = world.get_mut::<C>(entity).and_then(|t| t.next()) {
```
```suggestion
                if let Some(next) = world.get::<C>(entity).and_then(|t| t.next()) {
```

NthTensor commented:
I resolved all the issues identified in the reviews and I was in the process of porting over the examples from `bevy_eventlistener` when I identified a rather footgun with how global observers (ones who don't target a specific list of entities) interact with bubbling; eg, they don't.

I'll be rewriting this a bit over the next week to fix this issue.

aevyrie commented:
Completely non-blocking - have you done any sort of performance profiling? I'm curious how this scales.

aevyrie commented on crates/bevy_ecs/src/observer/mod.rs:
```diff
@@ -197,30 +219,82 @@ impl Observers {
-            if let Some(map) = observers.entity_observers.get(&entity) {
-                map.iter().for_each(&mut trigger_observer);
+            let mut entity = entity;
+            loop {
```
Isn't it possible to get into an infinite loop if the traversal is cyclic? Do we need to add an upper bound to this? Is documentation on `Traversal` sufficient?

NthTensor replied:
It is totally possible to make an infinite loop. I've done some tweaking to the docs to make them clearer, and in general I am not expecting most users to implement Traversal themselves. Open to bike shedding.

NthTensor commented:
I have not yet profiled. I am expecting it to be somewhat slower. Will port over the eventlistener stress test once I fix the footguns.

NthTensor commented:
Alright, resolved feedback, improved docs, changed how things worked a bit to work with global observers, simplified things a bit, added an example. Should be ready for reviews.

brandon-reinhart reviewed (approved):
I've followed this a bit. Looks good. Clean code, easy to understand. Good tests and example. Implementation seems flexible to future bevy features like relations and existing crates that provide types of relations (moonshine_view, for example).

Most UI event bubbling is just child-parent, but this is very general ECS event bubbling that can traverse the entity graph in whatever way the trait implementor wants. Super cool. The idea that the same system can carry the load of a click event or a damage event is great.

torsteingrindvik commented on crates/bevy_ecs/src/observer/mod.rs:
```diff
@@ -56,6 +62,34 @@ impl<'w, E, B: Bundle> Trigger<'w, E, B> {
+
+    /// Enables or disables event propagation, allowing the same event to trigger observers on a chain of different entities.
+    ///
+    /// The path an event will propagate along is specified by it's associated [`Traversal`] component. By default, events
```
```suggestion
    /// The path an event will propagate along is specified by its associated [`Traversal`] component. By default, events
```

torsteingrindvik commented on crates/bevy_ecs/src/observer/mod.rs:
```diff
@@ -662,4 +725,210 @@ mod tests {
+        // TODO: ideally this flush is not necessary, but right now observe() returns WorldEntityMut
+        // and therefore does not automatically flush.
+        world.flush();
+        world.trigger_targets(EventPropagating, [child, parent]);
```
Could you add a `world.trigger_targets(EventPropagating, [child, child]);` too?

I can't see docs about not being allowed to trigger the same entity twice so I'm curious what happens.
And I think I'd expect 4 on `R`?

NthTensor replied:
Will add. Should be allowed, would expect 4 as well.

torsteingrindvik commented on examples/ecs/observer_propagation.rs:
```diff
@@ -0,0 +1,126 @@
+        .run();
+}
+
+// In this example, we spawn a goblin wearing different pecies of armor. Each piece of armor
```
```suggestion
// In this example, we spawn a goblin wearing different pieces of armor. Each piece of armor
```

torsteingrindvik commented on examples/ecs/observer_propagation.rs:
```diff
@@ -0,0 +1,126 @@
+// In this example, we spawn a goblin wearing different pecies of armor. Each piece of armor
+// is represented as a child entity, with an `Armor` component.
+//
+// We're going to model how attack damage can be partally blocked by the goblin's armor using
```
```suggestion
// We're going to model how attack damage can be partially blocked by the goblin's armor using
```

torsteingrindvik commented on examples/ecs/observer_propagation.rs:
```diff
@@ -0,0 +1,126 @@
+// is represented as a child entity, with an `Armor` component.
+//
+// We're going to model how attack damage can be partally blocked by the goblin's armor using
+// event bubling. Our events will target the armor, and if the armor isn't strong enough to block
```
```suggestion
// event bubbling. Our events will target the armor, and if the armor isn't strong enough to block
```

torsteingrindvik commented on examples/ecs/observer_propagation.rs:
```diff
@@ -0,0 +1,126 @@
+        });
+}
+
+// This event represents an attack we want to "bubble" up from the armor to the goblin .
```
```suggestion
// This event represents an attack we want to "bubble" up from the armor to the goblin.
```

torsteingrindvik commented on examples/ecs/observer_propagation.rs:
```diff
@@ -0,0 +1,126 @@
+impl Event for Attack {
+    // 1. Which component we want to propagate along. In this case, we want to "bubble" (meaning propagate
+    //    from child to parent) so we use the `Parent` component for propagation. The component supplied
+    //    must implement the `Traversal` type.
```
```suggestion
    //    must implement the `Traversal` trait.
```

torsteingrindvik commented:
Since the trigger [entity](http://dev-docs.bevyengine.org/bevy/ecs/observer/struct.Trigger.html#method.entity) method typically returns the target entity of the explicit user API [trigger targets](http://dev-docs.bevyengine.org/bevy/ecs/prelude/struct.Commands.html#method.trigger_targets) I think it would be worth to add that propagation changes this implicitly when propagation is occurring. 
Imo a sentence about that in the entity method would be good

torsteingrindvik commented on crates/bevy_ecs/src/world/deferred_world.rs:
```diff
@@ -350,22 +351,49 @@ impl<'w> DeferredWorld<'w> {
+                self.reborrow(),
+                event,
+                entity,
+                components.iter().cloned(),
```
```suggestion
                components.iter().copied(),
```

alice-i-cecile commented on crates/bevy_ecs/src/event/base.rs:
```diff
@@ -31,7 +31,19 @@ use std::{
+pub trait Event: Component {
+    /// The component that describes which Entity to propagate this event to next, when [propagation] is enabled.
+    ///
+    /// [propagation]: crate::observers::Trigger::enable_propagation
```
```suggestion
    /// Defaults to [`TraverseNone`](crate::traverse::TraverseNone) when using the derive macro.
    /// This cannot be an associated type default as those are [still unstable](https://github.com/rust-lang/rust/issues/29661).
    /// [propagation]: crate::observers::Trigger::enable_propagation
```

alice-i-cecile reviewed (approved):
Good: very useful functionality, general but not-overdesigned, well-engineered. I left a comment to improve the docs and the extra test seems useful, but neither are blocking.

iiYese commented on crates/bevy_ecs/src/traversal.rs:
```diff
@@ -0,0 +1,39 @@
+/// [specify the direction]: crate::event::Event::Traverse
+/// [event propagation]: crate::observer::Trigger::propagate
+/// [observers]: crate::observer::Observer
+pub trait Traversal: Component {
```
So it will be highly likely that traversals will be along "relations at home" components. With hooks people can now ensure that these graphs never become invalid. However to do so underlying `Component` types have to be private otherwise you can break invariants by doing this:
```rs
let foo = world.entity(a).take::<Foo>().unwrap();
*world.entity(b).get_mut::<Foo>().unwrap() = foo;
```
I already do this in aery and hierarchy will have to do this when it starts using hooks so rather than being `Component` this should be `QueryData` and `next` should take the item instead of self. You can just remove `TraversalNone` and make the default query `()`.

iiYese replied:
Ugh actually this change might require #13375 first.

NthTensor replied:
Yep, I tried to design it this way originally and couldn't get the ref safety right, then designed a "minimal" version to use a single component. This would be best addressed as early follow-up, but I totally want this.

iiYese commented on crates/bevy_ecs/src/observer/mod.rs:
```diff
@@ -114,6 +148,15 @@ pub struct ObserverTrigger {
 
+/// Determines if the event should propagate to other entities.
+#[derive(Eq, PartialEq, Copy, Clone, Debug)]
+pub enum Propagation {
```
Is this enum necessary? I feel like this would be better as a `bool` because most of the time the thing that users will use to determine propagation is a boolean expression.

cart replied:
I agree. I think making this a bool, as a field that you directly access improves the legibility / usability of this (and removes the need for 3 separate accessor methods).

I'm putting together the change now for consideration.

cart replied:
Actually I'll take back that it should be a public field. In theory that would make it possible to override mutable borrow with a different one (although in practice this would be very challenging because it would need the same `w` lifetime). Better to keep that door closed though.

cart replied:
I've swapped this for a bool.

iiYese commented on crates/bevy_ecs/src/observer/mod.rs:
```diff
@@ -662,4 +725,210 @@ mod tests {
     }
+
+    #[test]
+    fn observer_propagating() {
```
Pedantic: I'm seeing many tests like this, not just for bubbling but observer in general. Tests using counters to determine if multiple observers ran or if a subset of observers ran. This isn't the best testing because the tests can pass when they shouldn't if an observer runs multiple times or if the wrong subset of observers run.

MiniaczQ replied:
simple fix is a `Vec<str>` where we can push observer "labels" when they execute and then verify with `vec!["obs1", "obs2"]`
it's only tests so it doesn't need to be "proper"

would that be alright?

NthTensor replied:
That seems correct to me. I'd rather not make this change as part of this PR. Changing the existing tests is out of scope, and I'd like to keep consistency. But I agree these tests could be improved as follow-up work.

iiYese commented on crates/bevy_ecs/src/observer/mod.rs:
```diff
@@ -18,15 +18,21 @@ use std::marker::PhantomData;
 /// [`Event`] data itself. If it was triggered for a specific [`Entity`], it includes that as well.
```
Should update the docs for this to say it includes propagation information.

cart replied:
Done!

iiYese reviewed (changes requested):
Few things I think should be changed but looks good overall.

cart reviewed (approved):
I like this impl! I agree with @iiYese that we should swap the enum for a bool. Feels cleaner given that it is the public interface we are presenting, and it allows us to simplify interacting with it.

I've pushed that change, alongside a rename from `Event::Traverse` to `Event::Traversal` (to align the concept name with the trait) and some doc additions / fixes.

MiniaczQ commented on crates/bevy_ecs/src/traversal.rs:
```diff
@@ -0,0 +1,39 @@
+/// events use the [`TraverseNone`] placeholder component, which cannot actually be created or added to
+/// an entity and so never causes traversal.
+///
+/// The implementer is responsible for ensuring that `Traversal::next` cannot produce infinite loops.
```
This isn't actually invalid is it? You can have an observer that triggers itself until it decides to stop propagation.
I'd keep it as a warning but not as a contract/responsibility.

NthTensor replied:
That's a good point. I'll add some more discussion about this.

MiniaczQ commented on crates/bevy_ecs/src/traversal.rs:
```diff
@@ -0,0 +1,39 @@
+
+impl Traversal for TraverseNone {
+    #[inline(always)]
+    fn next(&self) -> Option<Entity> {
```
Not a fan of `next` for non-iterator, for a solid minute I thought propagation can be sent to multiple targets

NthTensor replied:
Not sure how I feel about this. I'll think about other names.

cart replied:
Maybe just `traverse()`?

MiniaczQ reviewed (commented):
Very elegant :)

NthTensor commented:
Great, thank you! I prefer the bool as well, that's how I had it implemented before the first round of feedback.

I'll get this cleaned up and ready for merge by next monday then.

NthTensor commented on crates/bevy_ecs/src/observer/mod.rs:
```diff
@@ -56,6 +59,27 @@ impl<'w, E, B: Bundle> Trigger<'w, E, B> {
+        *self.propagate = should_propagate;
+    }
+
+    /// Returns the value of the  flag that controls event propagation. See [`propagate`] for more information.
```
Note to self: Remove extra space in this doc line.

NthTensor commented on crates/bevy_ecs/src/observer/mod.rs:
```diff
@@ -662,4 +706,206 @@ mod tests {
+            .spawn(Parent(parent))
+            .observe(|trigger: Trigger<EventPropagating>, mut res: ResMut<R>| {
+                res.0 += 1;
+                *trigger.propagate = false;
```
Note to self: This is now using a non-public api. Should call `Trigger::propagate` instead.

alice-i-cecile commented on crates/bevy_ecs/src/observer/mod.rs:
```diff
@@ -77,6 +77,8 @@ impl<'w, E, B: Bundle> Trigger<'w, E, B> {
 
     /// Returns the value of the flag that controls event propagation. See [`propagate`] for more information.
+    ///
+    /// [`propigate`]: Trigger::propagate
```
```suggestion
    /// [`propagate`]: Trigger::propagate
```

NthTensor commented:
Addressed review comments, added a benchmark, cleaned things up a bit. 

Preliminary profiling indicates this is a regression compared to the best case of `bevy_eventlisterner` but (a) its unlikely to cause problems in production and (b) we expect to be able to improve perf at a later stage.

If this passes tests I think it's good for merge.

alice-i-cecile commented:
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1657 if you'd like to help out.
